File: js/medium/BinaryTreeLevelOrderTraversalII.js
/**
 * Definition for a binary tree node.
 */
class TreeNode {
    constructor(val, left, right) {
        this.val = (val === undefined ? 0 : val);
        this.left = (left === undefined ? null : left);
        this.right = (right === undefined ? null : right);
    }
}

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
    if (!root) return [];
    
    const ans = [];
    const queue = [root];
    let queueIndex = 0;
    
    while (queueIndex < queue.length) {
        const levelSize = queue.length - queueIndex;
        const level = [];
        
        for (let i = 0; i < levelSize; i++) {
            const cur = queue[queueIndex++];
            level.push(cur.val);
            
            if (cur.left) queue.push(cur.left);
            if (cur.right) queue.push(cur.right);
        }
        
        ans.push(level);
    }
    
    // Reverse the result to get bottom-up order
    return ans.reverse();
};

export { TreeNode, levelOrderBottom };

File: py/medium/BinaryTreeLevelOrderTraversalII.py
import collections
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        queue = collections.deque()
        queue.append(root)
        ans = []
        while queue:
            qlen = len(queue)
            level = []
            for _ in range(qlen):
                top = queue.popleft()
                level.append(top.val)
                if top.left:
                    queue.append(top.left)
                if top.right:
                    queue.append(top.right)
            ans.append(level)

        # Reverse the result to get bottom-up order
        return ans[::-1]
